src/game_player.cpp

```diff
@@ -181,6 +177,30 @@ void Game_Player::UpdateScroll(int old_x, int old_y) {
 	int new_panx = new_x - screen_x;
 	int new_pany = new_y - screen_y;
 
+	// Detect whether we crossed map boundary.
+	// We need to scale down dx/dy to a single step
+	// to not message up further calculations.
+	// FIXME: This logic will break if something moves so fast
+	// as to cross half the map in 1 frame.
+	if (Game_Map::LoopHorizontal()) {
+		auto w = Game_Map::GetWidth() * SCREEN_TILE_WIDTH;
+		if (std::abs(dx) > w / 2) {
+			dx = (w - std::abs(dx)) % w;
+			if (new_x > old_x) {
+				dx = -dx;
+			}
+		}
+	}
+	if (Game_Map::LoopVertical()) {
+		auto h = Game_Map::GetHeight() * SCREEN_TILE_WIDTH;
```

Comment:
`SCREEN_TILE_HEIGHT`?

Reply:
It's misleading but there is no `SCREEN_TILE_HEIGHT`.

Reply:
Maybe we should generally rename it to `SCREEN_TILE_SIZE` then.

Reply:
Done